### ISSUE TITLE
RFE: Variable substitution in configuration

### ISSUE THREAD

**Issue by pboguslawski** [firewalld/firewalld#859]
**What would you like to be added**:

Variable substitution in firewalld items definitions similar to shorewalls params https://shorewall.org/manpages/shorewall-params.html

**Why is this needed**:

If you have many fw rules with same value i.e. IP address and then have to change this address (i.e. network readdressing), you need to change this value in many XML files. Any mistake can hurt. Better to change in one place (parameter definition file) and use parameter name in other places.

**Comment by erig0** [comment]
Can you use firewalld's `ipset` functionality? Those are usable for zone source and for rich rules.

**Comment by pboguslawski** [comment]
> Those are usable for zone source and for rich rules.

If ipset can be used everywhere in rules/policies (i.e. as dnat source/target address) without notable performance impact, it might be good solution and this issue may be closed but I can't see its possible in "destination address" nor "to-addr":

https://firewalld.org/documentation/man-pages/firewalld.policy.html

One more thing to consider for firewalld future: don' put any fw rules in zone definition and use separate item for it like "rule" in shorewall (seems that policies in firewalld are step in this direction but zone still allows defining rules which seems to be designed for local fw on host = not universal). Its worth to see shorewalls way of configuration fw stuff and borrow good solutions from it IHMO.

**Comment by erig0** [comment]
> > Those are usable for zone source and for rich rules.
> 
> If ipset can be used everywhere in rules/policies (i.e. as dnat source/target address) without notable performance impact, it might be good solution and this issue may be closed but I can't see its possible in "destination address" nor "to-addr":

Correct. It won't work for things that are designed to be single addresses, e.g. `to-addr`. I'll leave this RFE open. I think `destination` is easily addressed, but `to-addr` doesn't translate to an ipset well.

> https://firewalld.org/documentation/man-pages/firewalld.policy.html
> 
> One more thing to consider for firewalld future: don' put any fw rules in zone definition and use separate item for it like "rule" in shorewall (seems that policies in firewalld are step in this direction but zone still allows defining rules which seems to be designed for local fw on host = not universal). Its worth to see shorewalls way of configuration fw stuff and borrow good solutions from it IHMO.

I agree about zones. I'd like to see all rules attached to policies and zero rules attached to zones. Policies were just recently introduced, so we're on the right trajectory. I have more ideas about how to expand on policies so stay tuned. ;)
